dds/DCPS/RTPS/Sedp.cpp

```diff
@@ -4778,65 +4674,59 @@ void Sedp::resend_user_crypto_tokens(const RepoId& id)
+        get_handle_registry()->get_remote_datawriter_crypto_handle(*writer_pos);
+      if (dwch == DDS::HANDLE_NIL) {
+        continue;
       }
+      create_and_send_datareader_crypto_tokens(drch, reader_pos->first, dwch, *writer_pos);
     }
   }
 
   /*
    * For each user DataWriter that has a crypto handle, send tokens to matched
    * DataReaders that have a crypto handle.
    */
-  const DCPS::DatawriterCryptoHandleMap::iterator lwchm_end =
-    local_writer_crypto_handles_.end();
-  for (DCPS::DatawriterCryptoHandleMap::iterator lwchmi =
-        local_writer_crypto_handles_.begin();
-      lwchmi != lwchm_end; ++lwchmi) {
-    const RepoId& local_writer = lwchmi->first;
-    const DDS::Security::DatawriterCryptoHandle local_writer_ch = lwchmi->second;
-    if (DCPS::GuidConverter(local_writer).isUserDomainEntity()) {
-      const LocalPublicationIter lpi = local_publications_.find(local_writer);
-      if (lpi != local_publications_.end()) {
-        const DCPS::RepoIdSet::iterator me_end = lpi->second.matched_endpoints_.end();
-        for (DCPS::RepoIdSet::iterator mei = lpi->second.matched_endpoints_.begin();
-            mei != me_end; ++mei) {
-          const RepoId& remote_reader = *mei;
-          if (DCPS::GuidPrefixEqual()(remote_reader.guidPrefix, remote_participant.guidPrefix)) {
-            const DCPS::DatareaderCryptoHandleMap::iterator rdrchmi =
-              remote_reader_crypto_handles_.find(remote_reader);
-            if (rdrchmi != remote_reader_crypto_handles_.end()) {
-              const DDS::Security::DatareaderCryptoHandle remote_reader_ch = rdrchmi->second;
-              create_and_send_datawriter_crypto_tokens(
-                local_writer_ch, local_writer,
-                remote_reader_ch, remote_reader);
-            }
-          }
-        }
+  for (LocalPublicationCIter writer_pos = local_publications_.begin(),
+         writer_limit = local_publications_.end();
+       writer_pos != writer_limit; ++writer_pos) {
+    if (!DCPS::GuidConverter(writer_pos->first).isUserDomainEntity()) {
+      continue;
+    }
+    const DDS::Security::DatawriterCryptoHandle dwch =
+      get_handle_registry()->get_local_datawriter_crypto_handle(writer_pos->first);
+    if (dwch == DDS::HANDLE_NIL) {
+      continue;
+    }
+    for (DCPS::RepoIdSet::const_iterator reader_pos = writer_pos->second.matched_endpoints_.begin(),
+           reader_limit = writer_pos->second.matched_endpoints_.end();
+         reader_pos != reader_limit; ++reader_pos) {
+      if (!DCPS::GuidPrefixEqual()(reader_pos->guidPrefix, remote_participant.guidPrefix)) {
+        continue;
+      }
+      const DDS::Security::DatareaderCryptoHandle drch =
+        get_handle_registry()->get_remote_datareader_crypto_handle(*reader_pos);
+      if (drch == DDS::HANDLE_NIL) {
+        continue;
       }
+      create_and_send_datawriter_crypto_tokens(dwch, writer_pos->first, drch, *reader_pos);
     }
```

Comment:
The whole function look better. It isn't two solid blocks of if and for loop text anymore.